File: two-sum-iv-input-is-a-bst.cpp
//Runtime: 36 ms
class Solution {
public:
    TreeNode* temp;
    bool find(TreeNode* root, int f)
    {
        if(!root)
            return 0;

        if(root->val == f)
            temp = root;

        return find(root->left, f) | find(root->right, f);
    }

    bool ans = 0;

    TreeNode* mainroot;
    void solve(TreeNode* root, int t)
    {
        if(!root)
            return;

        find(mainroot, t-root->val);
        if(temp!=NULL && temp!=root)
        {
            ans = 1;
            return;
        }

        temp = NULL;
        solve(root->left, t);
        solve(root->right, t);
    }

    bool findTarget(TreeNode* root, int k)
    {
        if(!root)
            return 0;
        mainroot = root;
        solve(root, k);
        return ans;
    }
};

File: two-sum-iv-input-is-a-bst.py
# Runtime: 145 ms
class Solution(object):
    global mainroot

    def find(self, root, f, avoid):
        if not root:
            return False

        if root.val == f and root != avoid:
            return True

        if root.val < f:
            return self.find(root.right, f, avoid)

        return self.find(root.left, f, avoid)

    def solve(self, root, k):
        if not root:
            return False

        if self.find(self.mainroot, k - root.val, root):
            return True

        return self.solve(root.left, k) or self.solve(root.right, k)

    def findTarget(self, root, k):
        self.mainroot = root
        return self.solve(root, k)
